crates/provider/src/heart.rs

```diff
@@ -467,6 +496,26 @@ impl<S> Heartbeat<S> {
         // Blocks without numbers are ignored, as they're not part of the chain.
         let Some(block_height) = &block.header.number else { return };
 
+        // Add the block the lookbehind.
+        // The value is chosen arbitrarily to not have a huge memory footprint but still
+        // catch most cases where user subscribes for an already mined transaction.
+        // Note that we expect provider to check whether transaction is already mined
+        // before subscribing, so here we only need to consider time before sending a notification
+        // and processing it.
+        const MAX_BLOCKS_TO_RETAIN: usize = 10;
+        if self.past_blocks.len() >= MAX_BLOCKS_TO_RETAIN {
+            self.past_blocks.pop_front();
+        }
+        if let Some((last_height, _)) = self.past_blocks.back().as_ref() {
+            // Check that the chain is continuous.
+            if *last_height + 1 != *block_height {
+                // We cannot do anything about it, but we may at least log it.
+                warn!(%block_height, last_height, "non-continuous chain");
+            }
+        }
+        self.past_blocks.push_back((*block_height, block.transactions.hashes().copied().collect()));
```

Comment:
this doesn't handle reorgs, right?
here we'd append blocks with an existing height?
and we could find a tx confirmed in a reorged block  in L473

so I think this should replace the block with same number if it exists

Reply:
@mattsse That's right!

However, turns out the issue with reorgs is much deeper.
Once reorg occurs, we must also drop all the existing subscriptions received at block >= `*block_height`.
This part is easy, but it doesn't end here.
`ChainStreamPoller` doesn't seem to account for reorgs at all, e.g. [here](https://github.com/alloy-rs/alloy/blob/main/crates/provider/src/chain.rs#L72) it always expects the chain to be continuous.
And to fix that we probably need to compare block tip against `next_yield` [here](https://github.com/alloy-rs/alloy/blob/main/crates/provider/src/chain.rs#L56), and if `block_number < next_yield`, we should do some kind of search for the last non-diverging block (so probably we need to poll not only for block number, but rather for block header to use root hashes for comparison).

There is also an issue with the poller stream sporadically ending on transient network error, but that's a separate problem.

I've pushed a WIP commit with reorg handling in the `Heartbeat` and a (very flaky due to issues above) test.
PTAL, and let me know how you want to proceed.

If it works, I suggest the following:
- I will polish the commit, so that heartbeat supports reorgs, but will remove the test for now.
- We will merge this PR as-is, and I will create an issue for reorg handling in `ChainStreamPoller`.
- I will create a separate pull request for said issue, and will restore the test for reorg detection. (I cannot fully commit that I will do it due to limited capacity; but given that I already did some research and the problem seems interesting, I'll try my best to deliver it this week).

Reply:
thanks,
suggestions sgtm!

Reply:
I've removed the parts that are not relevant for this PR and created [an issue](https://github.com/alloy-rs/alloy/issues/1005) for the remaining work.
PTAL.